### DISCUSSION TITLE
Zoom-Proctoring not active

### DISCUSSION THREAD

**Discussion by heinzkrebs** [SafeExamBrowser/seb-server#43] [General]
dear all,

i try to run the zoom-proctoring feature. i've entered correct zoom API-keys/secrets and SDK-keys/secrets for a demo Quiz (MOCKUP) but when i run SEB with that configuration file i get connected to a lms.mockup.com, but the icon in the right lower corner says "Fernüberwachung ist nicht aktiv" - why? is there any log where i can find out what is going wrong????

side note: up to now i'm running SEB-secured exams on beginners of studies using Big Blue Button (WebRTC) as monitoring system, which works perfectly for 99.5% of windows user, but none of my (few) Mac-Users (for known reasons). But i cannot ignore them, and thus would like to test the built in zoom-client now ....

My current environment:
SEB-Server 1.2.4-SNAPSHOT
SEB 3.3.1

ciao
  Heinz A. Krebs

**Comment by anhefti** [comment]
Hi

Please be aware that proctoring with SEB Server is still a experimental feature. 
It also seems that you installed the development version of the SEB Server (-SNAPSHOT) that has no guarantee to work properly. To install the newest stable version of SEB Server 1.2-stable please refer to the setup Repo: https://github.com/SafeExamBrowser/seb-server-setup and the [installation guide](https://seb-server-setup.readthedocs.io/en/latest/overview.html). We recommend to install the prod-dockerhub installation.

Then the LMS Setup MOCKUP is just for testing without a real LMS. If you want to have a LMS and a real exam you need to bind a LMS, either Moodle, Open edX, Ans or OLAT. 
If you have setup the Zoom proctoring on SEB Server make sure you also set the proctoring in the Exam Configuration that you have applied to your exam. 

![image](https://user-images.githubusercontent.com/39078777/144002068-fcebeb46-66b2-4c1e-90df-f0b6735a53a2.png)

If this is all correctly setup and you have downloaded a Connection Configuration for SEB to apply to this exam, you should see the incoming SEB connection in the monitoring view of the SEB Server for this exam and SEB Server should apply the SEB connection automatically to a proctoring room that is a Zoom meeting to collect all the incoming SEB connections.

![image](https://user-images.githubusercontent.com/39078777/144002689-6d96e423-32ae-48a3-abc4-905a8b22c0c0.png)

For more details on how to setup SEB Server, please refer to the [User Guide ](https://seb-server.readthedocs.io/en/latest/#)

**Reply by heinzkrebs** [reply]
dear andreas hefti,

thanks for your answer!

i already switched over to 1.2.5 testing/basic without success, and yes, i know zoom proctoring is experimental, but currently i'm just playing around to find out in which direction i can go in the future ...

i have to use MOCKUP, since i do not have a supported LMS (but i'm able to import courses up to now), but in my opinion the proctoring feature does not depend on the LMS, does it?

all configurations are set correctly (in my opinion) since in SEB i see the icon "Fernwartung" and in the "running exams" tab i notice the incomming request from the SEB client - but i do not get connected to zoom.

and thus my question is if there is any further logging/info where i can take a look what could go wrong .....

ciao
  H.

**Comment by anhefti** [comment]
dear Heinz

The SEB Server do not add a connecting SEB client to a proctoring room until the connection is fully established and SEB was able to download the Exam Config from SEB Server and also login to the LMS. 

If you can see the SEB Connection as "Active" in the monitoring view and you also see that the SEB connection was applied to a Proctoring Room "Room 1 (**1**/20)" for example then the SEB Server has sent a meeting join instruction to the SEB to join the meeting.

If you don't see the SEB connection as "Active" but as "Requesting" in the monitoring view then the SEB probably don't get the join instruction from the SEB Server. 

So unfortunately I think proctoring is currently not working without an established SEB connection that also implies a proper LMS login. Maybe we can change this in future releases of SEB Server. The ability to communicate with a student before LMS login to discus eventually login problems seems like a valuable use case to me.

As a workaround you might try to open a one-to-one Room for the connected SEB client by double-click on the SEB connection in the monitoring view. You will get to the detail view of this SEB connection then. There you can try to use the "Open Single Room" action to initiate a one to one meeting with the participant. This might work even without a "Active" connection (if the  "Open Single Room" action is available then, I'm not quit sure about this, you have to look and try).

You can find the SEB log files in {Unser-Home}\AppData\Local\SafeExamBrowser\Logs
and for the SEB Server, if you have a normal Docker Setup, you can use **docker logs {servicename}** within a command line tool

best
Andreas

**Reply by heinzkrebs** [reply]
dear andreas,

well, i'm afraid i'm still too naive about what the SEB server is expected to do and what the workflow is :-o
and the second problem is, that i'm running an opensource assessment system, which i'm trying to learn how to behave like a Open edX-LMS ...

so what i could achive up to now is:
* i can setup my LMS as Open edX LMS Type
* SEB Server looks up my exams and displays them within the "Exam Administration>LMS Exams Lookup" page
* i can add the exam to exam-table in the database, so that it shows up on the "Exam Administration>Exam" page
* i can view the exam and configure the procotring settings and so on
* finally i export the Exam211206_2021-11-30.seb

so if i now run the SEB using this seb-file the seb-client gets connected to the SEB server, and requests the settings for the exam (since the seb-file only contains the sebServerURL and the exam iD and some other keys but no URL to the exam itself)
* the server sends some information back to the SEB, including the URL of the exam
* SEB connects to my assessent system and starts the exam (after logging in)

so this works from my side, i can see the client in the SEB server, and i can see if someone raises the hand. BUT the status is still "Connection Requested" and, as said before, the is no connection to zoom

so is the SEB Client expecting some "special answer" from the LMS to start a zoom session, or is the SEB server waiting for some answer? i notice some 404 answers for /seb-openedx/api/v1/course/ixam/127/configuration/ in the web-log ....

ciao
  H.

**Comment by anhefti** [comment]
dear Heinz

> so this works from my side, i can see the client in the SEB server, and i can see if someone raises the hand. BUT the status is still "Connection Requested" and, as said before, the is no connection to zoom

The problem here might be that your LMS is not a real Open edX system as I'm understanding this right from your explanation? 
SEB Server is closely bound to an LMS that is supported by the SEB Server and its open API to query course data and also to its login process to get the session identifier from the LMS login to request some user-account (username) data from the LMS for the specified SEB connection. This is in one hand needed to show the students name in the monitoring as well as creating a proctoring access token (with the students name) for joining the proctoring meeting. 

SEB Server do not manage LMS user accounts but just request some User-Account details from a specific LMS with the given LMS login identifier provides from the LMS to the SEB on login and provided further from SEB to the SEB Server. 

In your case I assume that the SEB is not able to get the LMS session identifier on LMS login and therefore is not able to finish up the connection handshake with the SEB Sever. The connection is still on the status "Connection Requested" even if you are logged into the exam. I Think this is the last missing step in your setup and that is needed before SEB Server sends the meeting join instruction to the SEB.

Open edX usually sends a cookie to the SEB on login. This cookie contains the edx session identifier that is seeking be a connecting SEB to finish up the handshake with the SEB Server. If you are able to also mock-up this behavior for your mock-up LMS then you can try to send a cookie header in the LMS login response to SEB like: 

`Set-Cookie: edx-user-info="{\"username\":\"lms-account-username\"}" `

This might work for your setup but with no guaranties.

There is still the possibility to setup a proctoring environment just with SEB and Zoom/Jitsi without SEB Server. In this case you need to setup a Meeting and proper SEB configuration by yourself. If this is an alternative scenario for you, please ask @danschlet for a setup on MacOS and iOS devices or @dbuechel for a setup on windows devices.

best, andreas

**Reply by heinzkrebs** [reply]
dear andreas,

i will investigate the cookie-exchange lateron, but:

* your final sentence makes me much more happy! i think for my special case the SEB-server is some kind of overkill (at present times), so to hear that there could be a possibility to directly connect SEB client with zoom server sounds great to me!!!!!!!!!!!!!!!!

thanks for your great help (for non-standard szenarios)!

ciao
  H.

**Reply by heinzkrebs** [reply]
i now got steps further :)))

* setrawcookie("edx-user-info",'{"username":"myUserName"}', time()+60, "", "", true);
  did the job, thanks for the info
* i had to mock the /api/user/v1/accounts... as well :)))

now SEB-server connects zoom, but does not succeed:
"Your request to add a new Licensed user could not be approved at this time because your account has reached the permitted maximum number of 1 paying users. Please purchase additional licenses or contact the Zoom support team to provision additional users in your account."

ok, sounds clear to me, i currently have just one account, but:
* is there need for ONE license per Proctoring room, so if i wanna run an exam with 100 candidates i would need 100 licenses??
* or just ONE license per exam, because the proctoring rooms are breakout rooms in zoom?

ciao
  H.

**Reply by anhefti** [reply]
dear Heinz

For proctoring with Zoom you need at least access to Zooms REST API and therefore at least a Pro, Business or Enterprise account plan:
![image](https://user-images.githubusercontent.com/39078777/145171147-6cec54d1-4072-4b8d-a670-693d3088d97e.png)

With SEB Server, the number of simultaneous meetings depends on how you define the proctoring room sizes and how many participants you have for an exam. town-hall and one to one rooms are additional (ad-hoc) meetings that are created in addition just when they are opened. 

We have tested it with an [educational account plan](https://zoomgov.com/education), but I don't know how many licenses are within this plan. Best would be to ask Zoom support directly up on your needs.

**Reply by heinzkrebs** [reply]
dear andreas,

all account plans have a limited number of licenses, and thus it's interesting, how SEB manages users and meetings.

so for example: eg 50 students are doing their exam
* if i define a proctering room size of 60, then only ONE new user WITH ONE meeting will be created?
* if i define a proctering room size of 10, then 5 new user with 5 meetings will be created?
* or will there be one new user/meeting for every student, thus 50 new user+50 meetings?

ciao
  H.

**Reply by anhefti** [reply]
dear Heinz

The first two points are correct for the normal (collecting) proctoring rooms. Then if one opens the town-hall an additional user/meeting is used. And again one for every one-to-one room that is opened simultaneously. 

But the users that are created with the Zoom API to manage the rooms are created from the one licensed user (the one that created the API Key's and JWT) and I'm not quit sure if the users that are created for the rooms fell under licensed user terms. This has to be clarified within Zoom support. The users that manages the proctoring meetings are of type "custCreate" and have no normal login or account:

![image](https://user-images.githubusercontent.com/39078777/145178902-e83ff2e4-ad61-4026-8af8-e2e0accea32b.png)

**Reply by heinzkrebs** [reply]
dear andreas,

the answer from zoom is, that those users are created with user_info->type 2 = licensed, and thus a license has to be available - which i do not have, and so the error occurs.

if i switch the type to 1 = basic, then the user can be created, as well as the meeting, and SEB gets connected, everything is fine - but basic accounts can only host meetings with a maximum duration of 40 minutes. and our exams last 2 hours or more ....

**Comment by danschlet** [comment]
> side note: up to now i'm running SEB-secured exams on beginners of studies using Big Blue Button (WebRTC) as monitoring system, which works perfectly for 99.5% of windows user, but none of my (few) Mac-Users (for known reasons). But i cannot ignore them, and thus would like to test the built in zoom-client now ....

Why is your BBB solution not working for Mac users? Since SEB 3.0, SEB for macOS supports access to camera/microphone via WebRTC on macOS 11.1 Big Sur and newer. You just need to make sure to use correct settings, see its [release notes](https://github.com/SafeExamBrowser/seb-mac/releases/tag/3.0). In case you also need access to screen sharing, we're also not yet supporting this with Zoom.

**Reply by heinzkrebs** [reply]
it's hard to say, since everything is running remotely in the lockdown now, and i'm not familiar with mac-environment. but i've been told that the camera does not appear, so i have the feeling, that SEB does not get access to the camera/microphone ...
and some/many of them are running a lower version than 11.1

and one student told me he is using 11.5.2 (and may did already an upgrade to 12.0.1), but the browser agent is still "Intel Mac OS X 10_15_7" ??????

**Reply by heinzkrebs** [reply]
Well, the one who can read is in advantage :)))
my settings contained "sendBrowserExamKey=true", and after setting this to false, at lease some of my students can access BBB within SEB, the other still have an older version, i assume ..

**Comment by danschlet** [comment]
We currently don't have any capacity to document access to Zoom proctoring using manual settings. The final release version of SEB 3.2 for macOS will offer according setting options in its Preferences window, but we will only be able to work on that later in January. Then we will also have time to document these settings. Also using manual Zoom proctoring settings without SEB Server will require some small changes in code at least in the macOS version.
